src/algebra/order/rearrangement.lean

```diff
@@ -0,0 +1,332 @@
+  rw mem_map,
+  obtain hi | hi := eq_or_ne (σ i) i,
+  { refine ⟨_, λ h, ⟨i, h, hi⟩⟩,
+    rintro ⟨j, hj, h⟩,
+      rwa σ.injective (hi.trans h.symm) },
+  { refine iff_of_true ⟨σ⁻¹ i, hs $ λ h, hi _, perm.apply_inv_self _ _⟩ (hs hi),
+    convert congr_arg σ h; exact ( perm.apply_inv_self _ _).symm }
+end
+
+@[to_additive]
+lemma prod_reindex [comm_monoid α] (s : finset ι) (f : ι → α) {σ : perm ι}
+  (hs : {x | σ x ≠ x} ⊆ s) :
+  (∏ i in s, f (σ i)) = ∏ i in s, f i :=
+begin
+  convert (prod_map _ σ.to_embedding _).symm,
+  exact (s.map_perm hs).symm,
+end
+
+-- TODO: `to_additive` fails this one
+lemma prod_reindex' [comm_monoid α] (s : finset ι) (f : ι → ι → α) {σ : perm ι}
+  (hs : {x | σ x ≠ x} ⊆ s) :
+  (∏ i in s, f (σ i) i) = ∏ i in s, f i (σ⁻¹ i) :=
+begin
+  convert s.prod_reindex (λ i, f i (σ⁻¹ i)) hs,
+  ext,
+  rw perm.inv_apply_self,
+end
+
+lemma sum_reindex' [add_comm_monoid α] (s : finset ι) (f : ι → ι → α) {σ : perm ι}
+  (hs : {x | σ x ≠ x} ⊆ s) :
+  (∑ i in s, f (σ i) i) = ∑ i in s, f i (σ⁻¹ i) :=
+begin
+  convert s.sum_reindex (λ i, f i (σ⁻¹ i)) hs,
+  ext,
+  rw perm.inv_apply_self,
+end
+
+end finset
+
+open finset equiv equiv.perm order_dual
+open_locale big_operators
+
+/-- **Rearrangement Inequality** : statement for a pair of functions that given a finset `s`,
+`movary_on f g s` holds with the permutation applied on the right.-/
+theorem monovary_on.sum_smul_comp_perm_le_sum_smul [linear_ordered_ring α]
+  [linear_ordered_add_comm_group β] [module α β] [ordered_smul α β]
+  {s : finset ι} {f : ι → α} {g : ι → β} (hfg : monovary_on f g s) {σ : perm ι}
+  (hσ : {x | σ x ≠ x} ⊆ s) :
+  ∑ i in s, f i • g (σ i) ≤ ∑ i in s, f i • g i :=
+begin
+  classical,
+  revert hσ σ hfg,
+  apply finset.induction_on_max_value₂ g f s,
+  { simp only [le_refl, finset.sum_empty, implies_true_iff] },
+  intros a s has hamax hind hfg σ hσ,
+  set k := σ a with hk,
+  set j := σ⁻¹ a with hj,
+  set p : ι → ι := λ x, if x = a then a else if x = j then k else σ x with hp,
+  set q : ι → ι := λ x, if x = a then a else if x = k then j else σ⁻¹ x with hq,
+  have hqpleft : function.left_inverse q p,
+  { intro x,
+    simp only [hp, hq],
+    split_ifs with h₁ h₂ h₃ h₄ h₅,
+    { rw h₁ },
+    { rw [h₂, hj, eq_inv_iff_eq, ← hk, h₃] },
+    { rw h₂ },
+    { exfalso,
+      apply h₂,
+      rw [hj, eq_inv_iff_eq, h₄] },
+    { exfalso,
+      apply h₁,
+      simp only [←inv_eq_iff_eq, inv_apply_self] at h₅,
+      exact h₅ },
+    { simp only [inv_apply_self] } },
+  have hqpright : function.right_inverse q p,
+  { intro x,
+    simp only [hp, hq],
+    split_ifs with h₁ h₂ h₃ h₄ h₅,
+    { rw ← h₁ },
+    { rw [h₂, hk, ← inv_eq_iff_eq, ← hj, h₃] },
+    { rw h₂ },
+    { exfalso,
+      apply h₂,
+      rw [hk, ← inv_eq_iff_eq, h₄] },
+    { exfalso,
+      apply h₁,
+      simp only [←inv_eq_iff_eq, inv_apply_self] at h₅,
+      exact h₅ },
+    { simp only [apply_inv_self] } },
+  set τ : perm ι := by use [p, q, hqpleft, hqpright] with hτ,
```

Comment:
I think there's an easier way of building this using `equiv.swap` and something about the subtype equal to `a`. Maybe @pechersky can remember where the pieces you'd need are.

Reply:
Golfed using, `equiv.swap`! Have a look if there's anything else you'd like to see golfed @eric-wieser